csrc/scheduler/matmul_utils.cpp

```diff
@@ -0,0 +1,795 @@
+    if (0 != (fusion_inputs.size() - fusion_inputs_tvs.size())) {
+      return {
+          c10::nullopt,
+          "Fusion inputs contain at least one non-TensorView object"};
+    }
+
+    // Fusion outputs have non-TensorView vals
+    if (0 != (fusion_outputs.size() - fusion_outputs_tvs.size())) {
+      return {
+          c10::nullopt,
+          "Fusion outputs contain at least one non-TensorView object"};
+    }
+
+    // Fusion input TensorViews have unsupported number of domains (2 for single
+    // gemm, 3 for batch gemms)
+    if (!areDimsValidInTvSet(fusion_inputs_tvs)) {
+      return {
+          c10::nullopt,
+          "One of fusion TensorView inputs has unsupported number of domains"};
+    }
+
+    // Fusion output TVs have unsupported number of domains (2 for a single
+    // gemm, 3 for batch gemms)
+    if (!areDimsValidInTvSet(fusion_outputs_tvs)) {
+      return {
+          c10::nullopt,
+          "One of fusion TensorView outputs has unsupported number of domains"};
+    }
+  }
+
+  // Expensive checks
+  {
+    // Get the number of how many times TensorView instances from provided set
+    // (e.g. fusion inputs / fusion outputs) are present in given paths
+    const auto getNumberOfTvInstancesInPaths =
+        [](const std::vector<Val*>& tv_instances,
+           const std::deque<std::deque<Val*>>& paths) {
+          size_t input_instances = 0;
+          for (const auto& path : paths) {
+            for (const auto* val : path) {
+              for (const auto* tv : tv_instances) {
+                input_instances += static_cast<size_t>(val->sameAs(tv));
+              }
+            }
+          }
+          return input_instances;
+        };
+
+    const auto paths_inputs_to_mma_A =
+        getAllDepndencyChains(fusion_inputs, {mma_inputs[0]});
+    const auto paths_inputs_to_mma_B =
+        getAllDepndencyChains(fusion_inputs, {mma_inputs[1]});
+    const auto paths_mma_C_to_outputs =
+        getAllDepndencyChains({mma_outputs[0]}, fusion_outputs);
+
+    // No paths from fusion inputs to MMA op input A
+    if (paths_inputs_to_mma_A.empty()) {
+      return {
+          c10::nullopt,
+          "There is no dependency between fusion inputs and MMA op first input"};
+    }
+
+    // No paths from fusion inputs to MMA op input B
+    if (paths_inputs_to_mma_B.empty()) {
+      return {
+          c10::nullopt,
+          "There is no dependency between fusion inputs and MMA op second input"};
+    }
+
+    // No paths from MMA op output to any of fusion outputs
+    if (paths_mma_C_to_outputs.empty()) {
+      return {
+          c10::nullopt,
+          "There is no dependency between MMA op output and fusion outputs"};
+    }
+
+    // The length of paths between MMA output and fusion outputs must 1 -
+    // this means there are no additional operations in between
+    {
+      for (const auto& path : paths_mma_C_to_outputs) {
+        if (path.size() != 1) {
```

Comment:
I am seeing you grabbing size 1 paths a lot. What's the benefit of grabbing such a path, compared with just the following?
```C++
if (!mma->output(0)->isFusionOutput()) {
  return {c10::nullopt, "One of paths between MMA op output and fusion outputs contains a product of unsupported operation"};
}
```
Are you thinking about potential fusion?